core/internal/runfiles/saved_file.go

```diff
@@ -141,12 +178,34 @@ func (f *savedFile) onFinishUpload(task *filetransfer.DefaultUploadTask) {
 	}
 
 	f.Lock()
+	defer f.Unlock()
+
+	// Record what we believe the server now has.
+	if task.Err == nil {
+		if f.uploadingB64MD5 != "" {
+			f.lastUploadedB64MD5 = f.uploadingB64MD5
+		} else if b64, err := hashencode.ComputeFileB64MD5(f.realPath); err == nil {
+			f.lastUploadedB64MD5 = b64
+		} else {
+			f.lastUploadedB64MD5 = ""
+		}
+	}
+
 	f.isUploading = false
 	if f.reuploadScheduled {
 		f.reuploadScheduled = false
-		f.doUpload(f.reuploadURL, f.reuploadHeaders)
+
+		// Unlock temporarily to compute hash.
+		f.Unlock()
+		currentB64, err := hashencode.ComputeFileB64MD5(f.realPath)
+		shouldReupload := (err != nil) || (f.lastUploadedB64MD5 == "") || (currentB64 != f.lastUploadedB64MD5)
+		f.uploadingB64MD5 = currentB64
+		f.Lock()
```

Comment:
Since we already compute a hash in `Upload`, why not store that? Your logic can be simpler with two fields instead:

* `lastUploadedB64MD5`
* `reuploadB64MD5` (similar to the other `reuploadXYZ` fields)

Instead of using an `uploadingB64MD5` field, let `doUpload` close over that information, similar to how `reuploadURL` and `reuploadHeaders` turn into its `uploadURL` and `uploadHeaders` arguments. It can then pass this `uploadB64MD5` parameter to `onFinishUpload` for the logic that updates `lastUploadB64MD5`.